guppylang/prelude/_internal/compiler/array.py

```diff
@@ -207,3 +225,30 @@ def compile_with_inouts(self, args: list[Wire]) -> CallReturnWires:
 
     def compile(self, args: list[Wire]) -> list[Wire]:
         raise InternalGuppyError("Call compile_with_inouts instead")
+
+
+class ArrayIterEndCompiler(ArrayCompiler):
+    """Compiler for the `ArrayIter.__end__` method."""
+
+    def compile(self, args: list[Wire]) -> list[Wire]:
+        # For linear array iterators, we need to make sure that all array elements have
+        # been taken out (i.e. replaced by `None`)
+        if self.elem_ty.type_bound() == ht.TypeBound.Any:
+            elem_opt_ty = ht.Option(self.elem_ty)
+            none_ty = ht.UnitSum(1)
+            # Define `unwrap_none` function
+            func = self.builder.define_function("unwrap_none", [elem_opt_ty], [none_ty])
+            err_msg = "Linear array element has not been used in iterator"
+            build_expect_none(
+                cast(DfBase[ops.DfParentOp], func), func.inputs()[0], err_msg
+            )
```

Comment:
Annoying that we have this `cast` here.

`func` is a `Function` which inherits `DfBase[ops.FuncDefn]` and `ops.FuncDefn` inherits `ops.DfParentOp`. The problem is that `DfBase` is not covariant, so this cast is technically not safe. But I also don't want to define `build_expect_none` only for `DfBase[ops.FuncDefn]`. Any ideas?

Reply:
See #676